src/Picasso_MarchingCubes.hpp

```diff
@@ -0,0 +1,648 @@
+        { 11, 0, 7 },  { 11, 1, 0 },  { 7, 0, 2 },   { 1, 5, 11 },
+        { 7, 9, 2 },   { 0, 4, 8 },   { 1, 5, 11 },  { 7, 9, 2 },
+        { 1, 10, 8 },  { 7, 5, 2 },   { 2, 5, 8 },   { 5, 1, 8 },
+        { 7, 5, 0 },   { 4, 1, 10 },  { 7, 0, 2 },   { 0, 10, 8 },
+        { 0, 1, 10 },  { 7, 9, 2 },   { 4, 1, 10 },  { 7, 9, 2 },
+        { 7, 5, 2 },   { 2, 5, 8 },   { 5, 4, 8 },   { 7, 5, 0 },
+        { 7, 0, 2 },   { 0, 4, 8 },   { 7, 9, 2 },   { 7, 9, 2 },
+        { 11, 10, 9 }, { 10, 6, 2 },  { 9, 10, 2 },  { 4, 0, 11 },
+        { 0, 9, 11 },  { 4, 11, 10 }, { 2, 8, 6 },   { 5, 11, 10 },
+        { 0, 5, 10 },  { 0, 6, 2 },   { 0, 10, 6 },  { 4, 11, 10 },
+        { 4, 5, 11 },  { 2, 8, 6 },   { 1, 4, 9 },   { 1, 9, 11 },
+        { 4, 6, 2 },   { 9, 4, 2 },   { 1, 0, 11 },  { 0, 9, 11 },
+        { 2, 8, 6 },   { 1, 5, 11 },  { 0, 4, 6 },   { 0, 6, 2 },
+        { 1, 5, 11 },  { 6, 2, 8 },   { 9, 1, 2 },   { 5, 1, 9 },
+        { 1, 6, 2 },   { 1, 10, 6 },  { 4, 1, 10 },  { 5, 0, 9 },
+        { 2, 8, 6 },   { 0, 1, 10 },  { 0, 6, 2 },   { 0, 10, 6 },
+        { 4, 1, 10 },  { 2, 8, 6 },   { 5, 4, 9 },   { 9, 4, 2 },
+        { 4, 6, 2 },   { 5, 0, 9 },   { 6, 2, 8 },   { 0, 4, 6 },
+        { 2, 0, 6 },   { 6, 2, 8 },   { 10, 8, 11 }, { 11, 8, 9 },
+        { 4, 0, 11 },  { 0, 9, 11 },  { 4, 11, 10 }, { 0, 5, 10 },
+        { 5, 11, 10 }, { 0, 10, 8 },  { 4, 5, 11 },  { 4, 11, 10 },
+        { 1, 4, 9 },   { 1, 9, 11 },  { 9, 4, 8 },   { 1, 0, 11 },
+        { 0, 9, 11 },  { 1, 5, 11 },  { 0, 4, 8 },   { 1, 5, 11 },
+        { 1, 10, 8 },  { 9, 5, 8 },   { 5, 1, 8 },   { 0, 9, 5 },
+        { 4, 1, 10 },  { 0, 1, 10 },  { 0, 10, 8 },  { 4, 1, 10 },
+        { 4, 9, 5 },   { 9, 4, 8 },   { 5, 0, 9 },   { 0, 4, 8 } };
+};
+
+//---------------------------------------------------------------------------//
+template <class MemorySpace>
+struct Data
+{
+    // Cell facet case ids and offsets. First value is the case id, second is
+    // the offset.
+    Kokkos::View<int*** [2], MemorySpace> cell_case_ids_and_offsets;
+
+    // Number of facets.
+    int num_facet;
+
+    // Facets. Facets are only created for owned cells.
+    Kokkos::View<double* [3][3], MemorySpace> facets;
+};
+
+//---------------------------------------------------------------------------//
+// Data creation funciton.
+template <class Mesh>
+Data<typename Mesh::memory_space> createData( const Mesh& mesh )
+{
+    auto cell_space = mesh.localGrid()->indexSpace(
+        Cajita::Ghost{}, Cajita::Cell{}, Cajita::Local{} );
+
+    Data<typename Mesh::memory_space> data;
+    data.cell_case_ids_and_offsets =
+        Kokkos::View<int*** [2], typename Mesh::memory_space>(
+            Kokkos::ViewAllocateWithoutInitializing(
+                "cell_case_ids_and_offset" ),
+            cell_space.extent( 0 ), cell_space.extent( 1 ),
+            cell_space.extent( 2 ) );
+    Kokkos::deep_copy( data.cell_case_ids_and_offsets, 0 );
```

Comment:
Why don't you let the constructor initialize then?

Reply:
good call - fixed. I still kept the `createData` function though for template deduction. It just calls the constructor now.